File: py/immpy/inmem_host_tree.py
from typing import Dict, List, Optional, TYPE_CHECKING, Callable
from dataclasses import dataclass

from mypy_utils import unwrap
from host_tree import HostNode
from node import Node, Properties

@dataclass
class InmemHostNodeData:
    kind: str
    properties: Properties

InmemHostNode = Node[InmemHostNodeData]

class InmemHostTree:
    def __init__(self):
        self.nodes: Dict[HostNode,InmemHostNode] = {}
        self.next_node_id = 0

    def begin_frame(self):
        pass

    def end_frame(self):
        pass

    def create_node(self, kind: str) -> HostNode:
        id = self._get_next_node_id()

        node = Node(InmemHostNodeData(kind=kind, properties={}))
        self.nodes[id] = node
        return id

    def destroy_node(self, node: HostNode):
        del self.nodes[node]

    def update_node_properties(self, node: HostNode, props: Properties):
        inmemnode = self.nodes[node]
        inmemnode.data.properties = props

    def remove_child_at(self, node: HostNode, index: int):
        inmemnode = self.nodes[node]
        inmemnode.remove_child_at(index)

    def insert_child_at(self, parent: HostNode, child: HostNode, index: int):
        inmemparent = self.nodes[parent]
        inmemchild = self.nodes[child]
        inmemparent.insert_child_at(inmemchild, index)

    def get_children(self, parent: HostNode) -> List[HostNode]:
        inmemparent = self.nodes[parent]
        inmem_children = inmemparent.children()
        return [unwrap(self._get_host_node(child)) for child in inmem_children]

    def trigger_event(self):
        pass

    def _get_host_node(self, child: InmemHostNode) -> Optional[HostNode]:
        for k in self.nodes:
            if self.nodes[k] == child:
                return k
        
        return None

    def _get_next_node_id(self) -> int:
        id = self.next_node_id
        self.next_node_id += 1

        return id


if TYPE_CHECKING:
    from host_tree import HostTree
    x: Callable[[],HostTree] = InmemHostTree

File: go/inmem_host_tree.go
package immgo

import (
	"fmt"
)

// InmemHostTree implements `HostTree` using a basic in-memory data structure.
//
// It is primarily intended to be used for testing.
type InmemHostTree struct {
	nodes map[HostNode]*Node[InmemHostNodeData]
	nextNodeId int
	root HostNode
}

type InmemHostNodeData struct {
	kind string
	properties Properties
}

func NewInmemHostTree() *InmemHostTree {
	ht := &InmemHostTree {
		nodes: make(map[HostNode]*Node[InmemHostNodeData]),
		nextNodeId: 0,
	}

	ht.root, _ = ht.CreateNode("root")
	return ht
}

func (this *InmemHostTree) GetRoot() HostNode {
	return this.root
}

func (this *InmemHostTree) BeginFrame() {
}

func (this *InmemHostTree) EndFrame() {
}

func (this *InmemHostTree) CreateNode(kind string) (HostNode, error) {
	id := this.getNextNodeId()

	node := NewNode[InmemHostNodeData]()
	node.data.kind = kind
	this.nodes[id] = node

	return id, nil
}

func (this *InmemHostTree) DestroyNode(node HostNode) error {
	_, exists := this.nodes[node]
	if !exists {
		return fmt.Errorf("Node does not exist")
	}

	delete(this.nodes, node)
	return nil
}

func (this *InmemHostTree) UpdateNodeProperties(node HostNode, props Properties) error {
	nodeRef, exists := this.nodes[node]
	if !exists {
		return fmt.Errorf("Node does not exist")
	}

	nodeRef.data.properties = props
	return nil
}

func (this *InmemHostTree) RemoveChildAt(parent HostNode, index int) error {
	nodeRef, exists := this.nodes[parent]
	if !exists {
		return fmt.Errorf("Node does not exist")
	}

	return nodeRef.RemoveChildAt(index)
}

func (this *InmemHostTree) InsertChildAt(parent HostNode, child HostNode, index int) error {
	parentRef, exists := this.nodes[parent]
	if !exists {
		return fmt.Errorf("Parent does not exist")
	}

	childRef, exists := this.nodes[child]
	if !exists {
		return fmt.Errorf("Child does not exist")
	}

	return parentRef.InsertChildAt(childRef, index)
}

func (this *InmemHostTree) GetChildren(parent HostNode) ([]HostNode, error) {
	parentRef, exists := this.nodes[parent]
	if !exists {
		return []HostNode{}, fmt.Errorf("Parent does not exist")
	}

	children := parentRef.Children()
	r := []HostNode{}
	for _, child := range children {
		r = append(r, this.GetHostNode(child))
	}

	return r, nil
}

func (this *InmemHostTree) TriggerEvent() error {
	// no-op (for now)
	return nil
}

func (this *InmemHostTree) GetNodeRef(hostNode HostNode) *Node[InmemHostNodeData] {
	return this.nodes[hostNode]
}

func (this *InmemHostTree) GetHostNode(node *Node[InmemHostNodeData]) HostNode {
	for k, v := range this.nodes {
		if v == node {
			return k
		}
	}


	return HostNode(-1)
}

func (this *InmemHostTree) getNextNodeId() HostNode {
	r := this.nextNodeId
	this.nextNodeId += 1
	return HostNode(r)
}
